sql: TestSchemaChangeEvalContext failed

[(sql).TestSchemaChangeEvalContext failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2322047&tab=buildLog) on [master@66a076f89e92becc1d7e1a2ecfd39c275165e2d8](https://github.com/cockroachdb/cockroach/commits/66a076f89e92becc1d7e1a2ecfd39c275165e2d8):

```
=== RUN   TestSchemaChangeEvalContext
--- FAIL: TestSchemaChangeEvalContext (2.49s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext982620575
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp;
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext982620575
    --- FAIL: TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp; (0.98s)
        schema_changer_test.go:3814: read the wrong number of rows: e = 1, v = 2

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSchemaChangeEvalContext PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSchemaChangeEvalContext.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestSchemaChangeEvalContext failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2322722&tab=buildLog) on [master@66a076f89e92becc1d7e1a2ecfd39c275165e2d8](https://github.com/cockroachdb/cockroach/commits/66a076f89e92becc1d7e1a2ecfd39c275165e2d8):

```
=== RUN   TestSchemaChangeEvalContext
--- FAIL: TestSchemaChangeEvalContext (1.21s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext678903418
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp;
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext678903418
    --- FAIL: TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp; (0.48s)
        schema_changer_test.go:3814: read the wrong number of rows: e = 1, v = 2

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestSchemaChangeEvalContext PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSchemaChangeEvalContext.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestSchemaChangeEvalContext failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2322723&tab=buildLog) on [master@66a076f89e92becc1d7e1a2ecfd39c275165e2d8](https://github.com/cockroachdb/cockroach/commits/66a076f89e92becc1d7e1a2ecfd39c275165e2d8):

```
=== RUN   TestSchemaChangeEvalContext
--- FAIL: TestSchemaChangeEvalContext (20.48s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext019750148
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp;
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext019750148
    --- FAIL: TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp; (8.60s)
        schema_changer_test.go:3814: read the wrong number of rows: e = 1, v = 2

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestSchemaChangeEvalContext PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSchemaChangeEvalContext.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestSchemaChangeEvalContext failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2324248&tab=buildLog) on [master@21f67e8534737f68ba9aad79aae8e24c1a6bdaba](https://github.com/cockroachdb/cockroach/commits/21f67e8534737f68ba9aad79aae8e24c1a6bdaba):

```
=== RUN   TestSchemaChangeEvalContext
--- FAIL: TestSchemaChangeEvalContext (22.35s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext175641954
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp;
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSchemaChangeEvalContext175641954
    --- FAIL: TestSchemaChangeEvalContext/ALTER_TABLE_t.test_ADD_COLUMN_x_TIMESTAMP_DEFAULT_current_timestamp; (10.04s)
        schema_changer_test.go:3814: read the wrong number of rows: e = 1, v = 2

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSchemaChangeEvalContext PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSchemaChangeEvalContext.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/backfill.go
+++ b/pkg/sql/backfill.go
@@ -956,6 +956,7 @@ func (sc *SchemaChanger) distBackfill(
 		// variable and assign to todoSpans after committing.
 		var updatedTodoSpans []roachpb.Span
 		if err := sc.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			updatedTodoSpans = todoSpans
 			// Report schema change progress. We define progress at this point
 			// as the the fraction of fully-backfilled ranges of the primary index of
 			// the table being scanned. Since we may have already modified the
@@ -988,7 +989,7 @@ func (sc *SchemaChanger) distBackfill(
 			}
 			metaFn := func(_ context.Context, meta *execinfrapb.ProducerMetadata) error {
 				if meta.BulkProcessorProgress != nil {
-					updatedTodoSpans = roachpb.SubtractSpans(todoSpans,
+					updatedTodoSpans = roachpb.SubtractSpans(updatedTodoSpans,
 						meta.BulkProcessorProgress.CompletedSpans)
 				}
 				return nil

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -3742,7 +3742,6 @@ CREATE TABLE d.t (
 // the backfilled columns.
 func TestSchemaChangeEvalContext(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 54775, "flaky test")
 	defer log.Scope(t).Close(t)
 	const numNodes = 3
 	const chunkSize = 200

